derive/src/optimizer/mod.rs

```diff
@@ -15,17 +16,30 @@ mod rotater;
 mod skipper;
 mod unroller;
 
+mod restorer;
+
 pub fn optimize(rules: Vec<Rule>) -> Vec<Rule> {
+    let rules_to_exprs = populate_rules_to_exprs(&rules);
+
     rules
         .into_iter()
         .map(rotater::rotate)
         .map(skipper::skip)
         .map(unroller::unroll)
         .map(concatenator::concatenate)
         .map(factorizer::factor)
+        .map(|rule| restorer::restore_on_err(rule, &rules_to_exprs))
         .collect()
 }
 
+fn populate_rules_to_exprs(rules: &[Rule]) -> HashMap<String, Expr> {
+    let mut rules_to_exprs = HashMap::new();
```

Comment:
There's a [nicer way](https://github.com/pest-parser/pest/blob/master/meta/src/validator.rs#L373) to write this. 😉 Some of that code could actually be extracted and made more abstract so that it can be used in both cases.